File: src/main/java/com/ploy/bubble_server_v3/domain/user/domain/vo/WashingRoom.java
package com.ploy.bubble_server_v3.domain.user.domain.vo;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor
public enum WashingRoom {
    // 열거형은 학기마다 달라질 수 있습니다.
    A41("A41", "A", 401, 417, 0, 0),
    A42("A42", "A", 418, 434, 0, 0),
    A31("A31", "A", 302, 307, 327, 334),
    A32("A32", "A", 308, 326, 0, 0),

    B41("B41", "B", 401, 417, 0, 0),
    B42("B42", "B", 418, 434, 0, 0),
    B31("B31", "B", 302, 307, 327, 334),
    B32("B32", "B", 308, 326, 0, 0),

    UNKNOWN("UNKNOWN", "", 0, 0, 0, 0);

    private final String code;
    private final String prefix;
    private final int minRoom;
    private final int maxRoom;
    private final int minRoomExtra;
    private final int maxRoomExtra;

    public static WashingRoom findWashingRoom(String prefix, int roomNumber) {
        for (WashingRoom room : values()) {
            if (room.prefix.equals(prefix) &&
                    ((room.minRoom <= roomNumber && roomNumber <= room.maxRoom) ||
                            (room.minRoomExtra <= roomNumber && roomNumber <= room.maxRoomExtra))) {
                return room;
            }
        }
        return UNKNOWN;
    }
}

File: src/main/kotlin/com/ploy/bubble_server_v2/domain/user/model/dto/vo/WashingRoom.kt
package com.ploy.bubble_server_v2.domain.user.model.dto.vo

import lombok.Getter
import lombok.RequiredArgsConstructor

data class RoomRange(val start: Int, val end: Int)

@Getter
@RequiredArgsConstructor
class WashingRoom private constructor(
    val name: String,
    val prefix: String,
    val roomRanges: List<RoomRange>
) {
    companion object {
        private val rooms = listOf(
            WashingRoom("B41", "B", listOf(RoomRange(401, 417))),
            WashingRoom("B42", "B", listOf(RoomRange(418, 434))),
            WashingRoom("B31", "B", listOf(RoomRange(302, 307), RoomRange(327, 334))),
            WashingRoom("B32", "B", listOf(RoomRange(308, 326)))
        )

        val UNKNOWN = WashingRoom("UNKNOWN", "", emptyList())

        fun findWashingRoom(prefix: String, roomNumber: Int): WashingRoom {
            return rooms.find { room ->
                room.prefix == prefix && room.roomRanges.any { range ->
                    roomNumber in range.start..range.end
                }
            } ?: UNKNOWN
        }

        fun values(): List<WashingRoom> = rooms + UNKNOWN
    }
}